File: test/noderegistry.lua
NodeProcess = ""

function setNodeProcess(processId)
  NodeProcess = processId
end

function addSuccessHandler(name, action)
  Handlers.add(
    name,
    Handlers.utils.hasMatchingTag("Action", action),
    function (msg)
      print("=========BEGIN=============")
      print("Action: " .. msg.Action)
      print("Message-Id: " .. msg["Message-Id"])
      print("Data: " .. msg.Data)
      print("=========END===============")
    end
  )
end

function addErrorHandler(name, action)
  Handlers.add(
    name,
    Handlers.utils.hasMatchingTag("Action", action),
    function (msg)
      print("=============BEGIN==========")
      print("Action: " .. msg.Action)
      print("Message-Id: " .. msg["Message-Id"])
      print("Error: " .. msg.Error)
      print("=============END============")
    end
  )
end

function addHandler(name, action)
  addSuccessHandler(name .. "Success", action .. "-Success")
  addErrorHandler(name .. "Error", action .. "-Error")
end

function testRegistry()
  addHandler("register", "Register")

  local action = "Register"
  local name = "aos"
  local publicKey = "Public Key"
  local desc = "Description"

  ao.send({Target = NodeProcess,  Action =  action})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey, Desc = desc})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey, Desc = desc})
end

function testUpdate()
  addHandler("update", "Update")

  local action = "update"
  local name = "aos"
  local name2 = "aos2"
  local publicKey = "Public Key"
  local desc = "Description"

  ao.send({Target = NodeProcess,  Action =  action})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name2})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey, Desc = desc})
  ao.send({Target = NodeProcess,  Action =  action, Name = name, Data = publicKey, Desc = desc})
end

function testDelete()
  addHandler("delete", "Delete")

  local action = "Delete"
  local name = "aos"
  local name2 = "aos2"
  local publicKey = "Public Key"
  local desc = "Description"

  ao.send({Target = NodeProcess,  Action =  action})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name2})
end

function testNodes()
  addHandler("nodes", "Nodes")

  local action = "Nodes"
  ao.send({Target = NodeProcess, Action = action})
end

function testGetNodeByName()
  addHandler("getNodeByName", "GetNodeByName")
  
  local action = "GetNodeByName"
  local name = "aos"
  local name = "aos2"

  ao.send({Target = NodeProcess,  Action =  action})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name})
  ao.send({Target = NodeProcess,  Action =  action, Name = name2})
end

function testAll()
  testRegistry()
  testUpdate()
  testDelete()
  testNodes()
  testGetNodeByName()
end

File: js_test/noderegistry.ts
import {message, result} from "@permaweb/aoconnect"
import {NODE_PROCESS} from "./constants"
import {getComputationProviderWallet, getMessage, ComputationProviderWallet} from "./utils"

export async function testAddWhiteList(address: string, computationProviderWallet: ComputationProviderWallet) {
    let action = "AddWhiteList"
    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
            {"name": "Address", "value": address},
        ]
    })

    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    })
    let Message = await getMessage(Result, msgId)
    return Message.Data
}

export async function testGetWhiteList(computationProviderWallet: ComputationProviderWallet) {
    let action = "GetWhiteList"
    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
        ],
    })

    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    })
    let Message = await getMessage(Result, msgId)
    return Message.Data
}

export async function testRemoveWhiteList(address: string, computationProviderWallet: ComputationProviderWallet) {
    let action = "RemoveWhiteList"
    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
            {"name": "Address", "value": address},
        ]
    })
    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    })
    let Message = await getMessage(Result, msgId)
    return Message.Data
}
async function testRegistry(name: string, computationProviderWallet:ComputationProviderWallet) {
    let action = "Register"
    let publicKey = "public key"
    let desc = Date() 

    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags" : [
            {"name": "Action", "value": action},
            {"name": "Name", "value": name},
            {"name": "Desc", "value": desc},
        ],
        "data": publicKey
    });

    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    });
    let Message = await getMessage(Result, msgId)
    return Message.Data
}
export async function testGetAllNodes(computationProviderWallet: ComputationProviderWallet) {
    let action = "Nodes"
    
    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
        ]
    });
    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    });
    let Message = await getMessage(Result, msgId)
    return Message.Data
}

async function testGetNodeByName(name: string, computationProviderWallet: ComputationProviderWallet) {
    let action = "GetNodeByName"

    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
            {"name": "Name", "value": name},
        ]
    });
    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    });
    let Message = await getMessage(Result, msgId)
    return Message.Data
}
async function testDelete(name: string, computationProviderWallet: ComputationProviderWallet) {
    let action = "Delete"

    let msgId = await message({
        "process": NODE_PROCESS,
        "signer": computationProviderWallet.signer,
        "tags": [
            {"name": "Action", "value": action},
            {"name": "Name", "value": name},
        ],
    });
    let Result = await result({
        "message": msgId,
        "process": NODE_PROCESS,
    });
    let Message = await getMessage(Result, msgId)
    return Message.Data
}

export async function registerAllNodes(names: string[], computationProviderWallet: ComputationProviderWallet) {
    for (let name of names) {
        let aos = await testRegistry(name, computationProviderWallet);
        console.log(`register ${name} result: ${aos}`)
    }
}

async function getAllNodesByName(names: string[], computationProviderWallet: ComputationProviderWallet) {
    for (let name of names) {
        let res = await testGetNodeByName(name, computationProviderWallet.signer)
        console.log(`get node by name ${name} result: ${res}`)
    }
}

export async function deleteAllNodes(names: string[], computationProviderWallet: ComputationProviderWallet) {
    for (let name of names) {
        let res = await testDelete(name, computationProviderWallet);
        console.log(`delete ${name} result: ${res}`)
    }
}

export async function main() {
    let computationProviderWallet = await getComputationProviderWallet();

    let nodes = ["js_aos", "js_aos2", "js_aos3"];
    await registerAllNodes(nodes, computationProviderWallet)

    let allProcess = await testGetAllNodes(computationProviderWallet);
    console.log("allProcess: ", allProcess)

    await getAllNodesByName(nodes, computationProviderWallet)

    await deleteAllNodes(nodes, computationProviderWallet)
}

// main()
